SMT2 floating point encoding fix

kroening opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [X] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

kroening committed:
implement cast bv -> floatbv

This case was missing but is used in our lowering of floating-point
operations.

kroening committed:
tolerate whitespace at end of SMT2 result

Z3, for instance, produces newlines at the end of the response; this should
be tolerated by the parser.

kroening commented:
This will need a non-recursive variant of letification, which is a few steps away from #4396.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 80e33fb21e22d6255cbee6a73af3e707bd294c1f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/104919955

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: cf71d76b988ef0c9e98ca59d7043334cb229c98e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/104925912

romainbrenguier commented on src/solvers/smt2/smt2irep.cpp:
```diff
@@ -47,7 +47,13 @@ bool smt2irept::parse()
-        throw error("unexpected end of file");
+        if(stack.empty())
+        {
+          result = irept(irep_idt());
```
Is this really necessary with the `result.clear()` above?

kroening replied:
No, not strictly; this is primarily meant to make it clear what is returned.

romainbrenguier replied:
I guess to really make it clear the best way would be to make `parse` return an `optionalt<irept>` instead of a bool.

tautschnig replied:
+1 for using `optionalt<irept>`

tautschnig replied:
But then also this commit should just be in a PR of its own where those kinds of changes could be made. It really doesn't relate to floating-point encoding fixes in any obvious way.

martin-cs commented on src/solvers/smt2/smt2_conv.cpp:
```diff
@@ -2316,6 +2316,20 @@ void smt2_convt::convert_typecast(const typecast_exprt &expr)
+        to_bv_type(src_type).get_width() !=
+        to_floatbv_type(dest_type).get_width())
+      {
+        UNEXPECTEDCASE("Typecast bv -> float with wrong width");
```
I think this is a `DATA_INVARIANT` of  typecast.  `UNHANDLED_CASE` was intended for `case` statements and as a route for converting old code that `case OTHER : assert(0);`.

tautschnig replied:
@kroening I believe this is still open.

kroening replied:
I don't think we guarantee that -- we do commonly cast between representations of different widths, say a cast from unsignedbv[8] to signedbv[32] is extremely common.

martin-cs replied:
Isn't that represented as a BV typecast though?

martin-cs replied:
We should be able to handle this by simply doing a BV typecast, then a BV to FP one.

kroening replied:
For this particular case, given that "bv" has no interpretation, it would probably be sensible to require that the width must match.
However, we don't have any of these at the moment, so this would be dead code either way.

martin-cs replied:
Fair.

martin-cs commented on src/solvers/smt2/smt2_conv.cpp:
```diff
@@ -2316,6 +2316,20 @@ void smt2_convt::convert_typecast(const typecast_exprt &expr)
+      }
+
+      if(use_FPA_theory)
+        SMT2_TODO("FPA cast bv -> float");
```
`((_ to_fp eb sb) (_ BitVec m) (_ FloatingPoint eb sb))` as given here http://smtlib.cs.uiowa.edu/theories-FloatingPoint.shtml

kroening committed:
smt2irep now returns optional

This is follow-up from a discussion in #4395 and clarifies the contract of
this function.

kroening commented:
There is now a separate commit with the optional return value of smt2irep.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: efebade97fc817bc802ee32fbe063e2fc115e12a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/105603393
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 9d2a61e4c2726c3f30c8cec8c268396d2c9e695b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/105604164
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 8c4c5e53296d7981fcd93995aeefcac8b2669a6e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/105604746
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: be64fc13ad01c494899b7c1597cedadbe93b8d24).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/105604892
